samples/nrf9160/https_client/src/main.c

```diff
@@ -0,0 +1,234 @@
+		 * with our security tag and reprovision our certificate.
+		 */
+		err = modem_key_mgmt_delete(TLS_SEC_TAG,
+					    MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN);
+		if (err) {
+			printk("Failed to delete existing certificate, err %d\n",
+			       err);
+		}
+	}
+
+	printk("Provisioning certificate\n");
+
+	/*  Provision certificate to the modem */
+	err = modem_key_mgmt_write(TLS_SEC_TAG,
+				   MODEM_KEY_MGMT_CRED_TYPE_CA_CHAIN,
+				   cert, sizeof(cert) - 1);
+	if (err) {
+		printk("Failed to provision certificate, err %d\n", err);
+		return err;
+	}
+
+	return 0;
+}
+
+/* Setup TLS options on a given socket */
+int tls_setup(int fd)
+{
+	int err;
+	int verify;
+
+	/* Security tag that we have provisioned the certificate with */
+	const sec_tag_t tls_sec_tag[] = {
+		TLS_SEC_TAG,
+	};
+
+	/* Set up TLS peer verification */
+	enum {
+		NONE = 0,
+		OPTIONAL = 1,
+		REQUIRED = 2,
+	};
+
+	verify = REQUIRED;
+
+	err = setsockopt(fd, SOL_TLS, TLS_PEER_VERIFY, &verify, sizeof(verify));
+	if (err) {
+		printk("Failed to setup peer verification, err %d\n", errno);
+		return err;
+	}
+
+	/* Associate the socket with the security tag
+	 * we have provisioned the certificate with.
+	 */
+	err = setsockopt(fd, SOL_TLS, TLS_SEC_TAG_LIST, tls_sec_tag,
+			 sizeof(tls_sec_tag));
+	if (err) {
+		printk("Failed to setup TLS sec tag, err %d\n", errno);
+		return err;
+	}
+
+	return 0;
+}
+
+void main(void)
+{
+	int fd;
+	int err;
+	int bytes;
+	struct addrinfo *res;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+	};
+	char *p;
+
+	printk("HTTPS client sample started\n\r");
+
+	err = bsdlib_init();
+	if (err) {
+		printk("Failed to initialize bsdlib!");
+		return;
+	}
+
+	/* Initialize AT comms in order to provision the certificate */
+	err = at_comms_init();
+	if (err) {
+		return;
+	}
+
+	/* Provision certificates before connecting to the LTE network */
+	err = cert_provision();
+	if (err) {
+		return;
+	}
+
+	printk("Waiting for network..\n");
+	err = lte_lc_init_and_connect();
+	if (err) {
+		printk("Failed to connect to the LTE network, err %d\n", err);
+		return;
+	}
+
+	err = getaddrinfo("google.com", NULL, &hints, &res);
+	if (err) {
+		printk("getaddrinfo() failed, err %d\n", errno);
+		return;
+	}
+
+	((struct sockaddr_in *)res->ai_addr)->sin_port = htons(HTTPS_PORT);
+
+	fd = socket(AF_INET, SOCK_STREAM, IPPROTO_TLS_1_2);
+	if (fd == -1) {
+		printk("Failed to open socket!\n");
+		goto clean_up;
+	}
+
+	/* Setup TLS socket options */
+	err = tls_setup(fd);
+	if (err) {
+		goto clean_up;
+	}
+
+	printk("Connecting to %s\n", "google.com");
+	err = connect(fd, res->ai_addr, sizeof(struct sockaddr_in));
+	if (err) {
+		printk("connect() failed, err: %d\n", errno);
+		goto clean_up;
+	}
+
+	bytes = send(fd, send_buf, strlen(send_buf), 0);
+	if (bytes < 0) {
+		printk("send() failed, err %d\n", errno);
+		goto clean_up;
+	}
+
+	bytes = recv(fd, recv_buf, RECV_BUF_SIZE, 0);
```

Comment:
We assume here and above that, we'll be able to `send`/`recv` data in a single shot. It keeps the code simple but is not exactly compliant with stream socket behavior. Perhaps we could provide `sendall`/`readall` helpers, to make sure we send/receive all data expected.

Reply:
Done

Reply:
What about `send`?

Reply:
Added :)